### PR TITLE
Added settings to disable individual notification types

### PR DESCRIPTION
**Pull request by aureumleonis**
# What
## Regarding the issue
Added the following settings/fields to users:
* `send_notifications_for_post_replies` (true by default)
* `send_notifications_for_comment_replies` (true by default)
* `send_notifications_for_private_messages` (true by default)
* `send_notifications_for_mentions` (true by default)

These are dependent on the "root" setting `send_notifications_to_email`, that is, if the individual settings are on but the root is off, no notifications are sent. If the root is on but no `false` value has been specified for a particular notification type, those should still be sent. This preserves backwards compatibility with current behavior.

The only change should be seen if the root is on but a given notification type is disabled. In that case all other notification types will be sent except for the disabled ones.

## Updates to test code
* Added a `test-utils` feature to db schema crate in order to expose a dev-only `TestDao` facility. This encapsulates boilerplate code needed to create/update users, instances, communities, posts and comments (so far)
* Added a `test-utils` feature to api commons to expose a simple test context factory

## Updates to lemmy context
The notification functions were tightly coupled with the `send_email` function. This made writing tests to verify email side effects extremely difficult. In fact, there were no tests for these side effects in the crate. In order to reduce coupling I introduced the `EmailSender` trait to be used with dynamic dispatch in the lemmy context. The default context constructor creates a sender pointing to the existing `send_email` function but allowing a different impl to be used in test code.

## Script updates
Added an optional second argument to `test.sh` to filter tests by module/test name, e.g.
```
$ ./scripts/test.sh lemmy_api_common build_response::tests::test_notifications
$ ./scripts/test.sh lemmy_api_common build_response::tests
```

## Migrations
Created a migration `2023-07-21-231007_add_granular_notification_settings` to create/drop columns for the new fields in the `local_user` table

# Why
Implements the backend for https://github.com/LemmyNet/lemmy/issues/1883

# Testing
Added tests to check notifications are sent when enabled and not sent when disabled.

# Notes
This was my first time setting up local dev for lemmy. The setup requires a relatively recent version of postgres (15+) which may catch you off guard.

### CODE REVIEW

**Comment by aureumleonis** [comment]
@dessalines @Nutomic @phiresky any comments?

File: crates/api_common/src/utils.rs [hunk 1]
```diff
@@ -307,25 +307,52 @@ pub fn honeypot_check(honeypot: &Option<String>) -> Result<(), LemmyError> {
...
   if let Some(user_email) = &local_user_view.local_user.email {
-    match send_email(
-      subject,
-      user_email,
-      &local_user_view.person.name,
-      body,
-      settings,
-    )
-    .await
+    match context
```
Comment (dessalines) [inline crates/api_common/src/utils.rs:346]: What's the point of adding an email_sender to the context? It isn't necessary, and works fine, with much less bloat, the way it was before.

You already did the `should_send` check, so everything else is unecessary.

File: crates/db_schema/src/test_utils.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
+#![allow(clippy::unwrap_used)]
```
Comment (dessalines) [inline crates/db_schema/src/test_utils.rs:1]: What does this have to do with this PR? Remove this.

File: crates/utils/src/email.rs [hunk 3]
```diff
@@ -19,6 +19,35 @@ pub mod translations {
 
 type AsyncSmtpTransport = lettre::AsyncSmtpTransport<lettre::Tokio1Executor>;
 
+#[async_trait::async_trait]
```
Comment (dessalines) [inline crates/utils/src/email.rs:22]: None of this is necessary.

File: crates/utils/Cargo.toml [hunk 4]
```diff
@@ -50,6 +50,7 @@ markdown-it = "0.5.1"
 totp-rs = { version = "5.0.2", features = ["gen_secret", "otpauth"] }
 ts-rs = { workspace = true, optional = true }
 enum-map = "2.6"
+async-trait = { workspace = true }
```
Comment (dessalines) [inline crates/utils/Cargo.toml:53]: Probably can be removed.

File: crates/api_common/src/test_utils.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#![allow(clippy::unwrap_used)]
```
Comment (dessalines) [inline crates/api_common/src/test_utils.rs:1]: Remove this, its unecessary.

File: crates/api_common/src/context.rs [hunk 6]
```diff
@@ -23,12 +25,23 @@ impl LemmyContext {
     client: ClientWithMiddleware,
     secret: Secret,
     rate_limit_cell: RateLimitCell,
+  ) -> LemmyContext {
+    let email_sender = Arc::new(DefaultEmailSender {});
```
Comment (dessalines) [inline crates/api_common/src/context.rs:29]: Not needed.

File: crates/api_common/src/build_response.rs [hunk 7]
```diff
@@ -217,3 +227,225 @@ pub async fn send_local_notifs(
 
   Ok(recipient_ids)
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (dessalines) [inline crates/api_common/src/build_response.rs:232]: You can probably remove all this, it just adds to code complication. Email and notification tests are pointless unless you can observe the recipient side.

**Review by dessalines — CHANGES_REQUESTED** [review]
Thanks! These will be useful in the future when we add proper UnifiedPush notifs as well.

There's a lot of cruft in this PR, you could probably cut it down 50% in size by removing the pointless email in context, as well as the tests that don't actually test this.

File: crates/api_common/src/utils.rs (hunk 1 above)
Comment (aureumleonis) [reply crates/api_common/src/utils.rs:346]: As explained under the "test" section in the PR description, there's currently no unit tests for the user preferences -> email handling logic. The "bloat" here is necessary to test this contract.

File: crates/db_schema/src/test_utils.rs (hunk 2 above)
Comment (aureumleonis) [reply crates/db_schema/src/test_utils.rs:1]: Clippy complains otherwise. This is test code but I think the `unwrap` uses could be just changed to `expect`

File: crates/utils/src/email.rs (hunk 3 above)
Comment (aureumleonis) [reply crates/utils/src/email.rs:22]: Discussed in https://github.com/LemmyNet/lemmy/pull/3742#discussion_r1282285277

File: scripts/test.sh [hunk 8]
```diff
@@ -5,8 +5,27 @@ CWD="$(cd -P -- "$(dirname -- "${BASH_SOURCE[0]}")" && pwd -P)"
 
 cd $CWD/../
 
+CMD_ARGS=
```
[the remark this answers is not in the document]
Comment (aureumleonis) [reply scripts/test.sh:8]: Care to explain why? This was also explained in the PR description, it allows execution of individual tests

File: crates/utils/Cargo.toml (hunk 4 above)
Comment (aureumleonis) [reply crates/utils/Cargo.toml:53]: discussed in https://github.com/LemmyNet/lemmy/pull/3742#discussion_r1282285277 for the overall test strategy

File: crates/api_common/src/test_utils.rs (hunk 5 above)
Comment (aureumleonis) [reply crates/api_common/src/test_utils.rs:1]: Will be addressed as per https://github.com/LemmyNet/lemmy/pull/3742#discussion_r1282137989

File: crates/api_common/src/context.rs (hunk 6 above)
Comment (aureumleonis) [reply crates/api_common/src/context.rs:29]: Discussed in https://github.com/LemmyNet/lemmy/pull/3742#discussion_r1282285277 for the overall test strategy

File: crates/api_common/src/build_response.rs (hunk 7 above)
Comment (aureumleonis) [reply crates/api_common/src/build_response.rs:232]: Strongly disagree. Not all tests need to or should be end to end. When projects grow these become ever more costly. On the other hand unit tests at the contract level keep things easy to refactor and safeguard future changes.

Comment (Nutomic) [reply crates/api_common/src/build_response.rs:232]: Testing is good, but it shouldnt be too much. In this case it looks like some enterprise software with all the abstractions. Its too complex and will make future changes unnecessarily complicated. Rust already ensures that refactoring doesnt break things.

Comment (dessalines) [reply crates/api_common/src/build_response.rs:232]: Agree. Rust already ensures in-code contracts. The unit tests we *do* have, are all testing the business-logic related to the database.

We don't need to do the enterprise java thing of creating DAOs and testing every tiny detail.

Comment (aureumleonis) [reply crates/api_common/src/build_response.rs:232]: As you mentioned here, only unit tests will help with bugs in logic like e.g. using an unintended field for a check or messing up a boolean condition. The logic to map certain user fields to behavior is precisely a logical thing not at all encoded in rust types. IMHO this isn't a "tiny detail" but a major feature. In fact, I know I could have done a quick and dirty job and simply modified the core logic in 10% of time it took to write the tests. I'm not a dogmatic person, e.g. I'd rather not dismiss "enterprise bloat" for philosophical reasons. Instead, let's take a look at concrete pros and cons. Though as you mention, y'all are the maintainers and deserve the last word.

Cons:
* `send_email` is now a dynamic call, a runtime change that's trivial compared to the time it takes to actually push bytes over the wire to send the data
* Additional field in lemmy context/god object. Set and forget. Email contract hasn't changed in decades, not expected to change soon
* Tests need to change if email/notification is refactored. Possible but unlikely, the tests are only testing the contract outcomes and _not_ any intermediate private logic
Pros:
* I was initially afraid of touching this code since it had _no_ tests checking contracts. Now it does and future contributors benefit from having their changes automatically verified. This means a smaller barrier on entry to making changes. Something I expect an OS project can benefit from
* Maintainers can quickly verify the extent of testing done just looking at the code, i.e. no need to remember minutiae of manual testing. This is specially useful for niche/less common used code paths
* Faster feedback on regressions. E.g. the tests encode proof that behavior is backwards compatible, and that should remain so for future changes

As mentioned before, it really is up to y'all but as a contributor I feel much safe updating code that already has thorough automated tests. As a code maintainer it should also make reviews much simpler.

Comment (dessalines) [reply crates/api_common/src/build_response.rs:232]: > Additional field in lemmy context/god object. Set and forget. Email contract hasn't changed in decades, not expected to change soon

The context should stay as small as possible, because its passed around a lot, all the apub calls, all the http calls, and injected in a lot of places. Nothing but your unit tests actually need an email in the context. That should be a red flag that none of it is necessary, and adding it to the context is unecessary bloat.

And 2ndly, this doesn't make the code simpler, just needlessly more complex. The only test worth doing there, is if you were to set up some kind of internal email service where you can verify that the email were actually sent, and the receiving side matches it.

So yes, please remove all the unecessary testing and just make this the settings addition and simple logical flag for when to send.

Comment (dessalines) [reply crates/api_common/src/build_response.rs:232]: Bump

**Comment by dessalines** [comment]
Stale PR. Can re-open if necessary, but should probably wait on the notification rework anyway.
